### PR TITLE
Lees-Edwards / lattice Boltzmann

### PR DESCRIPTION
**Pull request by bindgens1**
Fixes #

Description of changes:
We added the Lees-Edwards capability to the lattice Boltzmann GPU code (currently: two point coupling). Test cases are provided for  (i) transient velocity profile, (ii) boundary interpolation scheme for several cases and (iii) particle fluid coupling at the boundary. Documentation is already updated. 

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

**Comment by KaiSzuttor** [comment]
please check that it at least builds

File: src/core/lbgpu_cuda.cu [hunk 1]
```diff
@@ -613,6 +622,21 @@ __device__ void update_rho_v(float *mode, unsigned int index, LB_node_force_gpu
   d_v[index].v[2]=u_tot[2]; 
 }
 
+__device__ inline void equilibrium_modes(int ii, float rho, float *u, float *modes_from_pi_eq){
+
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please add a description

File: src/core/lbgpu_cuda.cu [hunk 2]
```diff
@@ -1018,7 +1032,299 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
   }
 }
 
+__device__ void calc_n_from_modes_push_LE(LB_nodes_gpu n_front, LB_nodes_gpu n_back, LB_nodes_gpu nodes_LE_upper, LB_nodes_gpu nodes_LE_lower, int index){
+
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please add a description

File: src/core/lbgpu_cuda.cu [hunk 3]
```diff
@@ -284,6 +284,15 @@ __device__ T xyz_to_index(T* xyz){
   return x + para.dim_x*(y + para.dim_y*z);
 }
 
+template < typename T >
+__device__ void index_to_xyz_LE(T index, T* xyz){
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: if possible dont use ouput variables

File: src/core/lbgpu_cuda.cu [hunk 4]
```diff
@@ -1018,7 +1032,299 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
...
+                     - mode[ 5 + ii * LBQ] - mode[ 6 + ii * LBQ]
+                     - mode[ 9 + ii * LBQ] + mode[11 + ii * LBQ]
+                     - mode[12 + ii * LBQ] - mode[14 + ii * LBQ]
+                     + mode[15 + ii * LBQ] + mode[16 + ii * LBQ]
+                     - mode[17 + ii * LBQ] - mode[18 + ii * LBQ]
+                   );
+ } 
+}
+
+__global__ void apply_LE_position_offset(LB_nodes_gpu n_front, LB_nodes_gpu n_back, float lees_edwards_offset) {
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: this function does a lot at once, maybe split it into subfunctions

File: testsuite/lb_gpu_leesedwards_velocity_profile.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+      # Compare results
+      quad_dev = 0.0
+      dev = U - x_vel
+
+      for i in np.arange(0, len(x_vel)):
+        quad_dev += dev[i] ** 2
+
+      self.assertTrue(quad_dev < tol)
+
+      # Integrate
```
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: superfluous comment

File: testsuite/lb_gpu_leesedwards_velocity_profile.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    # Add LB fluid
+    lb = md.lb.LBFluidGPU(agrid=1.0, dens=rho, visc=eta, fric=1.1, tau=time_step)
+    system.actors.add(lb)
+
+    X = np.arange(0, box_l) + 0.5
+    x_vel = np.empty(box_l)
+
+    while system.time <= total_time:
+
+      # Compute analystical solution
```
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: typo

**Comment by bindgens1** [comment]
Will improve this and check with the other myconfigs. Lees-Edwards doesn't play nice with other features.

File: testsuite/lb_gpu_leesedwards_velocity_profile.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import numpy as np
+import unittest as ut
+import espressomd as md
+from espressomd import lb
+
+
+@ut.skipIf(not md.has_features(['LB_GPU', 'LEES_EDWARDS']),
+  'Features not available, skipping test!')
+
+
```
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: remove newlines

File: testsuite/lb_gpu_leesedwards_particlecoupling.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+  def test(self):
+
+    """The velocity components at all LB nodes are tested against a table
+    stored in testsuite/data/"""
+
+    # Systemclass
+    system = md.System()
+
+    # allowed deviation
```
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_particlecoupling.py]: if you call it tolerance or similar youndont need the comment

File: testsuite/lb_gpu_leesedwards_gridoffset.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python
```
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_gridoffset.py]: please combine all the tests into one testclass

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: testsuite/lb_gpu_leesedwards_gridoffset.py (hunk 9 above)
Comment (bindgens1) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: How can I reset the LB to initial state?

Comment (fweik) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: You can't, as LB can not be disabled.

File: src/core/lbgpu_cuda.cu [hunk 10]
```diff
@@ -1815,7 +2122,7 @@ __global__ void temperature(LB_nodes_gpu n_a, float *cpu_jsquared, int *number_o
  * @param *delta                Pointer for the weighting of particle position (Output)
  * @param *interpolated_u       Pointer to the interpolated velocity (Output)
 */
-__device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, float *particle_position, unsigned int* node_index, float* mode, LB_rho_v_gpu *d_v, float* delta, float *interpolated_u) {
+__device__ __inline__ void interpolation_two_point_coupling( LB_nodes_gpu n_a, float *particle_position, unsigned int* node_index, float* mode, LB_rho_v_gpu *d_v, float* delta, float lees_edwards_offset, float le_position, float *interpolated_u ) {
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: the interpolation should really not depend on any LE variable

File: src/core/lbgpu_cuda.cu [hunk 11]
```diff
@@ -1968,7 +2344,7 @@ __device__ void calc_viscous_force(LB_nodes_gpu n_a, float *delta, float * partg
 #endif
 
   // Do the velocity interpolation
-  interpolation_two_point_coupling(n_a, position, node_index, mode, d_v, delta, interpolated_u);
+  interpolation_two_point_coupling(n_a, position, node_index, mode, d_v, delta, lees_edwards_offset, le_position, interpolated_u);
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: same as above, this function should not depend on LE variables

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: testsuite/lb_gpu_leesedwards_gridoffset.py (hunk 9 above)
Comment (KaiSzuttor) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: There is an open PR #1574 which should be merged and then used @fweik

File: src/core/lbgpu_cuda.cu (hunk 11 above)
Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I could provide two functions (the old and the new one for LE) and distinguish their call with an #ifdef / #ifndef. But that would nearly duplicate the functions.

Comment (KaiSzuttor) [reply src/core/lbgpu_cuda.cu]: No, just split functionality. Create a new function for LE and if it shares a lot of the two_point_coupling function, split that part off into a new function and call that new function from the two_point coupling and your function.

**Comment by bindgens1** [comment]
Clang on gitlab doesn't configure: 

`-- Configuring incomplete, errors occurred!
See also "/builds/espressomd/espresso/build/CMakeFiles/CMakeOutput.log".
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDART_LIBRARY`

Will merge the test cases next

**Comment by mkuron** [comment]
Just merge the current python branch to resolve the CUDART issue:
```
git pull upstream python
```

**Comment by bindgens1** [comment]
Thanks!

File: src/core/lbgpu_cuda.cu [hunk 12]
```diff
@@ -613,6 +622,26 @@ __device__ void update_rho_v(float *mode, unsigned int index, LB_node_force_gpu
   d_v[index].v[2]=u_tot[2]; 
 }
 
+
+/** This function computes the equilibrium modes as needed by the LB and
+* and the LB-LeesEdwards implementation
+* equilibrium part of the stress modes (eq13 schiller)*/
+
+__device__ inline void equilibrium_modes(int ii, float rho, float *u, float *modes_from_pi_eq){
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: output parameters...

File: src/core/lbgpu_cuda.cu [hunk 13]
```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
 }
 
 
+/*-------------------------------------------------------*/
+/**backtransformation from modespace to desityspace and streaming with the push method using pbc
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: typo density

File: src/core/lbgpu_cuda.cu [hunk 14]
```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
 }
 
 
+/*-------------------------------------------------------*/
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: remove that line

File: src/core/lbgpu_cuda.cu [hunk 15]
```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
...
+/**backtransformation from modespace to desityspace and streaming with the push method using pbc
+ * This time for the boundary nodes of the Lees-Edwards layers. 
+ * @param *nodes_LE_upper, nodes_LE_lower: Nodes of the upper and lowe LE boudary
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: typo lower, please describe each parameter

File: src/core/lbgpu_cuda.cu [hunk 16]
```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
...
+    mode[13 + ii * LBQ] = 0.0f;
+    mode[14 + ii * LBQ] = 0.0f;
+    mode[15 + ii * LBQ] = 0.0f;
+    mode[16 + ii * LBQ] = 0.0f;
+    mode[17 + ii * LBQ] = 0.0f;
+    mode[18 + ii * LBQ] = 0.0f;
+
+    //TODO later write into n_a and atomic_add n_a shifted and interpolated onto n_b
+    //TODO the sqrt around the weights 1/18 and 1/36 needs justification
+    n_back.vd[(4 + ii*LBQ ) * para.number_of_nodes + x
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: what happens here?

File: src/core/lbgpu_cuda.cu [hunk 17]
```diff
@@ -1803,19 +2126,9 @@ __global__ void temperature(LB_nodes_gpu n_a, float *cpu_jsquared, int *number_o
...
- * @param n_a                   Pointer to local node residing in array a (Input)
- * @param *particle_position    Pointer to the particle position (Input)
- * @param node_index            node index around (8) particle (Output)
- * @param *mode                 Pointer to the 19 modes for current lattice point (Output)
- * @param *d_v                  Pointer to local device values
- * @param *delta                Pointer for the weighting of particle position (Output)
- * @param *interpolated_u       Pointer to the interpolated velocity (Output)
-*/
-__device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, float *particle_position, unsigned int* node_index, float* mode, LB_rho_v_gpu *d_v, float* delta, float *interpolated_u) {
+// Calculating the xyz-position of the left node a particle
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please name your functions according to what they do, i.e. calculate.... Please add doxygen documentation

File: src/core/lbgpu_cuda.cu [hunk 18]
```diff
@@ -1843,9 +2156,13 @@ __device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, fl
...
-  int x = left_node_index[0] + para.dim_x;
-  int y = left_node_index[1] + para.dim_y;
-  int z = left_node_index[2] + para.dim_z;
+  *x = left_node_index[0] + para.dim_x;
+  *y = left_node_index[1] + para.dim_y;
+  *z = left_node_index[2] + para.dim_z;
+}
+
+// The nearest 8 nodes for every particle is calculated based on the xyz-postion of the left node of a particle 
+__device__ __inline__ void calculation_nearest_nodes(int x, int y, int z, LB_nodes_gpu n_a, unsigned int* node_index) {
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: Same as above, doxygen documentation and name of function should be verb.

File: src/core/lbgpu_cuda.cu [hunk 19]
```diff
@@ -1857,6 +2174,30 @@ __device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, fl
...
+/**(Eq. (12) Ahlrichs and Duenweg, JCP 111(17):8225 (1999))
+ * @param n_a                   Pointer to local node residing in array a (Input)
+ * @param *particle_position    Pointer to the particle position (Input)
+ * @param node_index            node index around (8) particle (Output)
+ * @param *mode                 Pointer to the 19 modes for current lattice point (Output)
+ * @param *d_v                  Pointer to local device values
+ * @param *delta                Pointer for the weighting of particle position (Output)
+ * @param *interpolated_u       Pointer to the interpolated velocity (Output)
+*/
+
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: remove empty line

File: src/core/lbgpu_cuda.cu [hunk 20]
```diff
@@ -1857,6 +2174,30 @@ __device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, fl
...
+__device__ __inline__ void interpolation_two_point_coupling( LB_nodes_gpu n_a, float *particle_position, unsigned int* node_index, float* mode, LB_rho_v_gpu *d_v, float* delta, float *interpolated_u ) {
+  
+  int x;
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: int x, y, z;

File: src/core/lbgpu_cuda.cu (hunk 18 above)
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: Please don't use output variables

File: src/core/lbgpu_cuda.cu [hunk 21]
```diff
@@ -1911,7 +2304,7 @@ __device__ __inline__ void interpolation_two_point_coupling(LB_nodes_gpu n_a, fl
  * @param *d_v                  Pointer to local device values
  * @param flag_cs               Determine if we are at the centre (0, typical) or at the source (1, swimmer only)
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu:2362]: doxygen documentation missing

File: src/core/lbgpu_cuda.cu [hunk 22]
```diff
@@ -1968,7 +2373,13 @@ __device__ void calc_viscous_force(LB_nodes_gpu n_a, float *delta, float * partg
 #endif
 
   // Do the velocity interpolation
+  
+#ifndef LEES_EDWARDS  
   interpolation_two_point_coupling(n_a, position, node_index, mode, d_v, delta, interpolated_u);
+#endif
+#ifdef LEES_EDWARDS
+  interpolation_two_point_coupling_LE(n_a, position, node_index, mode, d_v, delta, lees_edwards_offset, le_position, interpolated_u);
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: cant you call the specifics to your algorithm from within the interpolation function and decide if it's called there?

File: src/core/lbgpu_cuda.cu [hunk 23]
```diff
@@ -2826,6 +3237,101 @@ __global__ void reset_boundaries(LB_nodes_gpu n_a, LB_nodes_gpu n_b){
     n_a.boundary[index] = n_b.boundary[index] = 0;
 }
 
+__device__ void calculate_LE_mode_delta(int index, LB_rho_v_gpu *d_v, float lees_edwards_velocity, LB_nodes_gpu nodes_LE_upper, LB_nodes_gpu nodes_LE_lower){
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: doxygen documentation missing

Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: don't use output variables

File: src/core/lbgpu_cuda.cu [hunk 24]
```diff
@@ -2826,6 +3237,101 @@ __global__ void reset_boundaries(LB_nodes_gpu n_a, LB_nodes_gpu n_b){
...
+    delta_m[14*para.dim_x*para.dim_z + delta_index]  = 0.0f;
+    delta_m[15*para.dim_x*para.dim_z + delta_index]  = 0.0f;
+    delta_m[16*para.dim_x*para.dim_z + delta_index]  = 0.0f;
+    delta_m[17*para.dim_x*para.dim_z + delta_index]  = 0.0f;
+    delta_m[18*para.dim_x*para.dim_z + delta_index]  = 0.0f;
+  
+  }
+}
+
+__global__ void apply_LE_velocity_shift(LB_nodes_gpu n_front, LB_nodes_gpu n_back, LB_nodes_gpu nodes_LE_upper, LB_nodes_gpu nodes_LE_lower, LB_rho_v_gpu *d_v, float lees_edwards_velocity) {
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: doxygen documentation missing

File: src/core/lbgpu_cuda.cu [hunk 25]
```diff
@@ -2849,15 +3354,15 @@ __global__ void integrate(LB_nodes_gpu n_a, LB_nodes_gpu n_b, LB_rho_v_gpu *d_v,
...
+    relax_modes(mode, index, node_f, d_v);
     /**lb_thermalize_modes */
     if (para.fluct)
     {
       thermalize_modes(mode, index, &rng);
     }
 #if  defined(EXTERNAL_FORCES)  ||   defined (SHANCHEN)  
     /**if external force is used apply node force */
-    apply_forces(index, mode, node_f,d_v);
+apply_forces(index, mode, node_f,d_v);
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: clang-format -style=file -i

File: src/core/lbgpu_cuda.cu [hunk 26]
```diff
@@ -3683,15 +4195,19 @@ void lb_integrate_GPU() {
   /**call of fluid step*/
   /* NOTE: if pi is needed at every integration step, one should call an extended version 
            of the integrate kernel, or pass also device_rho_v_pi and make sure that either 
-           it or device_rho_v are nullptr depending on extended_values_flag */ 
+           it or device_rho_v are NULL depending on extended_values_flag */
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: nullptr was correct

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: src/core/lbgpu_cuda.cu [hunk 27]
```diff
@@ -1019,6 +1038,309 @@ __device__ void calc_n_from_modes_push(LB_nodes_gpu n_b, float *mode, unsigned i
...
+    mode[12 + ii * LBQ] = 0.0f;
+    mode[13 + ii * LBQ] = 0.0f;
+    mode[14 + ii * LBQ] = 0.0f;
+    mode[15 + ii * LBQ] = 0.0f;
+    mode[16 + ii * LBQ] = 0.0f;
+    mode[17 + ii * LBQ] = 0.0f;
+    mode[18 + ii * LBQ] = 0.0f;
+
+    //TODO later write into n_a and atomic_add n_a shifted and interpolated onto n_b
+    //TODO the sqrt around the weights 1/18 and 1/36 needs justification
```
Comment (uschille) [inline src/core/lbgpu_cuda.cu]: I suspect the `sqrt`s are wrong. The back transformation - see Eq. (2.66) in http://pubman.mpdl.mpg.de/pubman/item/escidoc:1423983/component/escidoc:1423982/Dissertation-Schiller.pdf needs to include the norm of the basis vectors which I do not see here. Cf. the statement `m[i] = (1./e[19][i])*mode[i];` in `lb_calc_n_from_modes(...)`. The quadrature weights `w[i]` should then enter without the `sqrt`.

Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I tried and it works much better. Will include that in the next commit! Thanks!

**Comment by bindgens1** [comment]
I will remove the prints and put all test cases in one file (with the box_l provided a the right spot) during one of the next commits.

File: build-lble/testsuite/lb_leesedwards.py [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+
+# LB parameter for test against Navier-Stokes equation
+eta = 0.27
+rho = 0.97
+nu = eta / rho
+v = 0.0087
+k_max = 100
+agrid = 1.0
+fric = 1.1
+lb2 = md.lb.LBFluidGPU(agrid=agrid, dens=rho, visc=eta, fric=fric, tau=time_step)
```
Comment (KaiSzuttor) [inline build-lble/testsuite/lb_leesedwards.py]: the gpu version is a feature that may not be compiled in, so you have to guard those statements that refer to the LBFluidGPU instance. Furthermore it's easier to set the system up in the class and only use a single testclass.

File: src/core/lbgpu_cuda.cu (hunk 3 above)
Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I don't really get what you mean. Are you referring to the xyz?

Comment (KaiSzuttor) [reply src/core/lbgpu_cuda.cu]: yes

File: src/core/lbgpu_cuda.cu [hunk 29]
```diff
@@ -292,13 +292,13 @@ __device__ T xyz_to_index(T* xyz){
 }
 
 template < typename T >
-__device__ void index_to_xyz_LE(T index, T* xyz){
-  xyz[0] = index%para.dim_x;
+__device__ void index_to_xyz_LE(T index, T* xyz_LE){
```
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: i meant to not use output variable at all and instead returning the value

**Comment by KaiSzuttor** [comment]
any news here?

**Comment by bindgens1** [comment]
As soon as the regular LE is merged I'll update this branch.

**Comment by fweik** [comment]
@bindgens1 my understanding is that we do not go forward with this, and instead directly into walberla, so this can be closed?

**Comment by bindgens1** [comment]
Sad but true :-/
If you want to remove the LB / LBGPU soon, it's not worth to put the effort in this project.

**Comment by fweik** [comment]
Sorry about that, so sad. As you might have noticed I'm also not happy with how this project progresses, let's just hope that we'll succeed eventually.

**Pull request closed without merging**